core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -857,6 +858,364 @@ KOKKOS_INLINE_FUNCTION_DELETED Impl::ThreadVectorRangeBoundariesStruct<
+  // args... is passed by value because it should always be ints
+  parallel_for(
+      nested_policy(status, policy.team, policy.boundaries[CurrentRank]),
+      [&](int const& i) {
+        if constexpr (Iter == Iterate::Right) {
+          nested_loop(typename NextStatus::type(), NextStatus(), policy, lambda,
+                      val, args..., i);
+        } else {
+          nested_loop(typename NextStatus::type(), NextStatus(), policy, lambda,
+                      val, i, args...);
+        }
+      });
+}
+
+template <MDTeamRangeThreadAndVector threadAndVector, typename MDTeamPolicy,
+          typename Lambda, typename ReductionValueType, typename Status>
+KOKKOS_INLINE_FUNCTION void md_parallel_impl(MDTeamPolicy const& policy,
+                                             Lambda const& lambda,
+                                             Status const& rangeStatus,
+                                             ReductionValueType& val) {
+  constexpr auto TotalRank = MDTeamPolicy::total_rank;
+  constexpr auto Iter      = MDTeamPolicy::iter;
+  using TeamHandle         = typename MDTeamPolicy::team_handle_type;
+
+  static_assert(MDTeamPolicy::total_rank >= 2 && MDTeamPolicy::total_rank <= 8);
+
+  using par_rank_t =
+      Impl::ParallelRank<TotalRank, Iter, typename TeamHandle::execution_space,
+                         threadAndVector>;
+
+  constexpr int startRank = (Iter == Iterate::Right) ? 0 : (TotalRank - 1);
+
+  using md_team_iter =
+      Impl::MDTeamRangeNextStatus<startRank, TotalRank, par_rank_t::par_rt,
+                                  par_rank_t::invalid, Iter>;
+
+  nested_loop(rangeStatus, md_team_iter(), policy, lambda, val);
+}
+
+}  // namespace Impl
+
+template <Iterate Iter, int TotalRank, typename TeamHandle>
+struct MDTeamThreadRange {
+  using rank_type        = int;
+  using boundary_type    = int;
+  using team_handle_type = TeamHandle;
+  using execution_space  = typename team_handle_type::execution_space;
+  using array_layout     = typename execution_space::array_layout;
+  static constexpr rank_type total_rank = TotalRank;
+  static constexpr Iterate iter         = Iter;
+
+  static constexpr Iterate direction =
+      iter == Iterate::Default
+          ? layout_iterate_type_selector<array_layout>::outer_iteration_pattern
+          : iter;
+
+  template <class... Args>
+  KOKKOS_INLINE_FUNCTION MDTeamThreadRange(team_handle_type const& team_,
+                                           Args&&... args)
+      : team(team_), boundaries{static_cast<boundary_type>(args)...} {
+    static_assert(sizeof...(Args) == TotalRank);
+  }
+
+  team_handle_type const& team;
+  boundary_type boundaries[total_rank];
+};
+
+template <typename TeamHandle, typename... Args>
+MDTeamThreadRange(TeamHandle const&, Args&&...)
+    ->MDTeamThreadRange<Kokkos::Iterate::Default, sizeof...(Args), TeamHandle>;
+
+template <Iterate Iter, int TotalRank, typename TeamHandle>
+struct MDThreadVectorRange {
```

Comment:
Same thing lets use Kokkos::Rank

Reply:
Changed in [c534937](https://github.com/kokkos/kokkos/pull/5238/commits/c5349378887bbbf53a29cd76eb21095cdd18b83d)